### ISSUE TITLE
Google calendar access disabled

### ISSUE THREAD

**Issue by szucskrisztian** [Zren/plasma-applet-eventcalendar#339]
When I try to connect to Gcal and click on the link in the widget settings, the browser opens and say that the request is invalid and I should contact with the developer. 

Issue details:
"400 hibakód: invalid_request
The out-of-band (OOB) flow has been blocked in order to keep users secure. Follow the Out-of-Band (OOB) flow migration guide linked in the developer docs below to migrate your app to an alternative method.
A kérelem adatai: redirect_uri=urn:ietf:wg:oauth:2.0:oob"

**Comment by szucskrisztian** [comment]
> r

And have you found any solution?

**Comment by mkwsnyder** [comment]
I have not; from the error message you got it sounds like Google changed how third party apps need to authenticate. Hopefully it won't be too difficult of a change, but without digging into the code and Google's documentation (something I can't do for a week or two) I'm not sure.

I even tried copying over the authentication token found in `~/.config/plasma-org.kde.plasma.desktop-appletsrc` from a machine I'd previously authenticated on and that didn't work.

**Comment by etoven** [comment]
> I have not; from the error message you got it sounds like Google changed how third party apps need to authenticate. Hopefully it won't be too difficult of a change, but without digging into the code and Google's documentation (something I can't do for a week or two) I'm not sure.
> 
> I even tried copying over the authentication token found in `~/.config/plasma-org.kde.plasma.desktop-appletsrc` from a machine I'd previously authenticated on and that didn't work.
> 
> **Edit:** it appears this is a duplicate issue of both #331 and #333 so I'd recommend closing it. [It looks like what I tried above is working for some people in said issues; I'd recommend trying it yourself to see if it works.](https://github.com/Zren/plasma-applet-eventcalendar/issues/331#issuecomment-1294960958)

That's not a fix if you cannot recover a token from a existing install.

**Comment by mkwsnyder** [comment]
> That's not a fix if you cannot recover a token from a existing install.

I agree, it's a temporary band-aid solution until the project owner can implement an actual fix. I'm simply mentioning it here so that it can help those that do have an existing install.

**Comment by FeryET** [comment]
Have the same issue. Is there anyway to genereate the token manually?

**Comment by JackDinn** [comment]
Please just fix. Iv tried all other calenders and they all have something that stops me from using them.

I wont be holding my breath.

**Comment by etoven** [comment]
He probably doesn't want to spin up a server or pay for api access anymore.

**Comment by cognitus** [comment]
@Zren is possible fix this?

**Comment by cruz94** [comment]
same issue here @Zren

**Comment by TaliesinEmrys** [comment]
Like many others, I'm experiencing the same problem.

It was mentioned in another thread that the code can be extracted from a config file and re-entered as a work around.

Which is good to know.

Unfortunately I did a completely fresh install, after some hardware upgrades.

Reformatted the old HDD and no longer have access to the file in question.

Which brings to mind a question, despite some recent changes in the last 4 months, (translation), is this applet actively maintained anymore?

Perhaps there is an active fork?

Anyone have any insights?

**Comment by tategotoazarasi** [comment]
Same issue.

**Comment by dan-ros** [comment]
The akonadi stuff can synchronize with google calendar, so maybe via that...

**Comment by Evgeny-del** [comment]
![скрин](https://user-images.githubusercontent.com/78655923/210577206-69e47df7-430c-42d6-a9bf-e69c56dba11f.png)

**Comment by deivid11** [comment]
Same issue here, cannot use google calendar sync

**Comment by mcasaccio** [comment]
same issue here
![image](https://user-images.githubusercontent.com/7736740/211632807-24bee69b-7724-427f-b1a7-64d7aa87c25a.png)

**Comment by tomholvey** [comment]
I'm also seeing this issue

**Comment by jackmonty** [comment]
I'm also seeing this issue.

**Comment by Sriramgireesh** [comment]
I am also facing this issue

**Comment by MoarcyVega** [comment]
I've found a way to show my events from google calendar:
It works also with the KDE-Pim-modul (, if you only need to see your events). 
The only thing that doesn't work this way: 
If you have a few calendars in your google calendar with different colors, the events don't get the color from your google calendar.

So i still hope for a fix.

**Comment by Evgeny-del** [comment]
https://github.com/Zren/plasma-applet-eventcalendar/issues/339#issuecomment-1396653697
Не могли бы Вы подробнее рассказать об этом способе, который Вы нашли?

**Comment by ruanpereira** [comment]
same issue, invalid request when clicking in the link

**Comment by ruanpereira** [comment]
you can do it with korganizer, but isn't the same thing

**Comment by HornetoMano** [comment]
> 

How you do this bro?

**Comment by ArTikOnesh** [comment]
> [#339 (comment)](https://github.com/Zren/plasma-applet-eventcalendar/issues/339#issuecomment-1396653697) Не могли бы Вы подробнее рассказать об этом способе, который Вы нашли?

Install kdepim-addons and kde kalendar. Import your google calendar into kde kalendar and after this import from widget like PIM events.

**Comment by ruanpereira** [comment]
just like @ArTikOnesh said. With KDE kalendar, go to the settings and add your google account with the events. Go back and activate the PIM events. If you use a KDE desktop environment, the PIM events can be activated in the configuration of the clock.

**Comment by Salvaju29ro** [comment]
> @HornetoMano proprio come@ArTikOneshdisse. Con KDE kalendar, vai alle impostazioni e aggiungi il tuo account Google con gli eventi. Torna indietro e attiva gli eventi PIM. Se si utilizza un ambiente desktop KDE, gli eventi PIM possono essere attivati ​​nella configurazione dell'orologio.

Is Google working again with Akonadi? I switched to Event Calendar because Kalendar didn't let me add events to Google (Korganizer also didn't let me)

**Comment by kanocz** [comment]
tried changed obb to localhost method... changed `redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob` to `redirect_uri=http%3A%2F%2F127.0.0.1:8080` and created small golang http server to get result
code:
`package main

import (
  "fmt"
  "net/http"
  "log"
  "html"
)

type httpDump struct {

}

func (httpDump) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()

	fmt.Fprintf(w, "Code: %s", r.FormValue("code"))


	fmt.Printf("URL, %q\n", html.EscapeString(r.URL.Path))
	fmt.Printf("Form: %+v\n", r.Form)
	fmt.Printf("Headers: %+v\n", r.Header)

}

func main() {

  log.Fatal(http.ListenAndServe(":8080", httpDump{}))

}`

It allows me to authorize, I'm getting code back... but after inserting in configuration text filed I'm getting this:
![image](https://user-images.githubusercontent.com/4784978/218719324-00d94e65-1236-43c2-aab9-0cf185bbe729.png)

anyone has better ideas? :)

**Comment by kanocz** [comment]
oh... it's 2-phase... after code we need to fetch access token... this part also have to be modified... will try to solve this later today

**Comment by kanocz** [comment]
Proof-of-concept idea works! :partying_face: With many manual steps for now, but.... I'm almost happy :)

![image](https://user-images.githubusercontent.com/4784978/218777619-a1d9bd18-ae7e-4cd5-a9aa-519c9c80fde3.png)

**Comment by materemias** [comment]
if you could create a MR it'd be great

**Comment by kanocz** [comment]
> if you could create a MR it'd be great

I'll try, but don't know how much time it'll take :( I need to read documentation about plasmoid development, is there any build-in http server in QML or if I can use some C++/go/rust code and connect it to QML script... or at least if it's possible to exec external app and get back it's output to automate all tasks I've done manually

**Comment by kanocz** [comment]
so... just finished temporary working solution... with little bit less manual steps :)

cli tool in golang
https://github.com/kanocz/plasma-applet-eventcalendar/blob/master/google-oauth-helper/local-http.go

just download it, and run as `go run local-http.go` (you need to have golang installed)
it will parse your event-plasmoid configuration, then generate URL and print it (in console)
copy-paste it (or CTRL+click if you use konsole)
then auth as always and then it will print config for plasmoid just in browser... you'll need to copy-paste it to your `.config/plasma-org.kde.plasma.desktop-appletsrc` and then probably logout-login

works for me, hope that will work for you also
P.S.: probably I'll integrate/automate it later, probably not :)

**Comment by bigo72** [comment]
> so... just finished temporary working solution... with little bit less manual steps :)
> 
> cli tool in golang https://github.com/kanocz/plasma-applet-eventcalendar/blob/master/google-oauth-helper/local-http.go
> 
> just download it, and run as `go run local-http.go` (you need to have golang installed) it will parse your event-plasmoid configuration, then generate URL and print it (in console) copy-paste it (or CTRL+click if you use konsole) then auth as always and then it will print config for plasmoid just in browser... you'll need to copy-paste it to your `.config/plasma-org.kde.plasma.desktop-appletsrc` and then probably logout-login
> 
> works for me, hope that will work for you also P.S.: probably I'll integrate/automate it later, probably not :)

@kanocz you're great! Anyway I've got an error by golang:
go run local-http.go
local-http.go:1:3: expected 'package', found 1
what it means? Am I making any mistake?

**Comment by kanocz** [comment]
> @kanocz you're great! Anyway I've got an error by golang: go run local-http.go local-http.go:1:3: expected 'package', found 1 what it means? Am I making any mistake?

it looks like you've copied file it with line numbers 😅 try to click on three dots and choose "view raw"

**Comment by bigo72** [comment]
> > @kanocz you're great! Anyway I've got an error by golang: go run local-http.go local-http.go:1:3: expected 'package', found 1 what it means? Am I making any mistake?
> 
> it looks like you've copied file it with line numbers sweat_smile try to click on three dots and choose "view raw"

Ohmygod, I'm so noob! Thank you!
It generated the link in Konsole but nothing else. I'm really sorry to bother you, but you are my only chance to solve this issue.
It doesn't print any configuration in browser

**Comment by bigo72** [comment]
![Screenshot_6](https://user-images.githubusercontent.com/6362701/218982116-c188f816-eac2-4e73-a2f3-f9f790534d07.png)
Here it is. Translation:

Access blocked: authorization error

bigo72[at]gmail.com
Missing required parameter: client_id Learn more about this error
If you are a developer of this app, see the details of the error.
Error 400: invalid_request

**Comment by rnc** [comment]
I have just tried this tool (although I still had a working accessToken) and it does correctly generate a new accessToken/accessTokenExpiresAt/accessTokenType and refreshToken values in the browser. I see it parses the existing `.config/plasma-org.kde.plasma.desktop-appletsrc` configuration looking for a `[Configuration][Google Calendar]` block - if I modify my existing configuration so that block doesn't exist then I can reproduce the above error. I am guessing @bigo72 does not have a correctly configured applet with a google calendar?

**Comment by kanocz** [comment]
yes, I need ClientToken and ClientSecret from existing config, so first you need open widget configuration and go to google calendar page... you don't need to click on URL there but at this moment widget will generate part of config I need to read

**Comment by rnc** [comment]
@kanocz It might be worth adding some error handling if `sessionClientSecret` and `sessionClientId` cannot be found in an existing config?

**Comment by kanocz** [comment]
or maybe I'll found some time today-tomorrow - will try

**Comment by miminno** [comment]
> yes, I need ClientToken and ClientSecret from existing config, so first you need open widget configuration and go to google calendar page... you don't need to click on URL there but at this moment widget will generate part of config I need to read

No matter how many times I try going to google calendar page in the settings, widget does not generate ClientToken and ClientSecret :( What's the secret?

**Comment by kanocz** [comment]
hm... strange... ok, I'll try to do better solution asap :) I've just found part of code how to execute external tool just from QML, so I'll pass clientID and clientSecret as arguments directly... and return config via stdout, so no copy-paste will be needed...

**Comment by kanocz** [comment]
ok.... https://github.com/Zren/plasma-applet-eventcalendar/pull/341
this fill just fix everything - no copy-paste needed and so on
P.S.: you need to have golang installed before `./install` executing, then you can remove it if you want

**Comment by lucafulgenzi** [comment]
@kanocz I just tried it and it works perfectly.
Excellent work 👍🏻

**Comment by bigo72** [comment]
I can't try now! I'm at work! I can't wait!
I'll let you know, but I'm pretty sure it will work. 
Thank you so much!

**Comment by tomholvey** [comment]
This worked for me too, thank you!

**Comment by miminno** [comment]
That works perfectly, thanks!

**Comment by kanocz** [comment]
I think it's ok - many deprecated APIs are used in this applet... just go to settings and this may work :)

**Comment by mkwsnyder** [comment]
@kanocz the current version currently errors out with the following (sorry it doesn't have more than this):
```sh
invalid params
exit status 2
```

If I use [yesterday's (2023-02-14) commit](https://github.com/kanocz/plasma-applet-eventcalendar/commit/f6a20d2a669789900168c178715b87720959b50c) it seems to run correctly.

Regardless, thank you for your work and rapid responses to people's requests!

**Comment by kanocz** [comment]
@mkwsnyder you don't need to run it manually anymore! just `git clone` of my version, then `./install` and use UI of the widget

**Comment by lucafulgenzi** [comment]
@bigo72 to be safe uninstall the old version first and then reinstalls from the fork, you don't need to run the script, but just go to the component configuration in the Google tab

**Comment by bigo72** [comment]
Thanks @lucafulgenzi , I did it before, but unfortunately it still doesn't work for me. I'm thinking ... does this script take auth from somewhere else? Is this "somewhere else" the "online accounts"?

**Comment by kanocz** [comment]
@bigo72 no... it just redirects you to Google oauth page and then read back some values, gets token and returns to widget which updates config in usual way... 
P.S.: if you want you can record screen and I'll try to help

**Comment by bigo72** [comment]
It worked!!!
I was still using the old version! I'm so sorry @kanocz 
I had to manually remove /usr/share/plasma/plasmoids/org.kde.plasma.eventcalendar/ directory, then I gitcloned your repo again and installed as usual (and this time without any error).
Thanks again for everything you did.

**Comment by kanocz** [comment]
sorry, one more commit added to pull request which is not connected to this issue... I just used master branch and forgot that's connected to pull request :(

P.S.: it's just configuration for different color for events in past, so I hope this doesn't metter

![image](https://user-images.githubusercontent.com/4784978/219611442-e1d712c1-83a3-4ae9-a314-288b32d479bc.png)

![image](https://user-images.githubusercontent.com/4784978/219611818-bc31e167-8dd6-49a0-9680-6718117e2cdc.png)

**Comment by felipelutz** [comment]
Thank you @kanocz. It worked flawlessly!

**Comment by podxboq** [comment]
> ok.... #341 this fill just fix everything

It works perfectly. You are my hero.

**Comment by TLPNull** [comment]
@kanocz thanks, I am able to login now, but I am unable to sync my calendar 
I enabled debugging and here are some logs

```
qml: [eventcalendar:debug] upcomingEvents.onAllDataFetched 2023-02-22T03:00:00.000Z 2023-03-08T16:02:32.167Z 2023-04-07T03:00:00.000Z
qml: HTTP 0 Headers:

file:///home/user/.local/share/plasma/plasmoids/org.kde.plasma.eventcalendar/contents/ui/config/GoogleLoginManager.qml:147: TypeError: Cannot call method 'debugJSON' of null
qml: XMLHttpRequest.onerror 0
file:///home/user/.local/share/plasma/plasmoids/org.kde.plasma.eventcalendar/contents/ui/config/GoogleLoginManager.qml:147: TypeError: Cannot call method 'debugJSON' of null
qml: HTTP 0 Headers:

qml: fetchGoogleTasklistList.response HTTP Error 0  0
file:///home/user/.local/share/plasma/plasmoids/org.kde.plasma.eventcalendar/contents/ui/config/GoogleLoginManager.qml:179: TypeError: Cannot call method 'debugJSON' of null
qml: XMLHttpRequest.onerror 0
qml: fetchGoogleTasklistList.response HTTP Error 0 null 0
file:///home/user/.local/share/plasma/plasmoids/org.kde.plasma.eventcalendar/contents/ui/config/GoogleLoginManager.qml:179: TypeError: Cannot call method 'debugJSON' of null
qml: [eventcalendar:debug] isExpanded true
qml: [eventcalendar:debug] setToday qua mar 8 13:03:10 2023 GMT-0300
qml: [eventcalendar:debug] currentTime qua mar 8 13:03:10 2023 GMT-0300
qml: [eventcalendar:debug] isExpanded false
qml: [logger:debug] updateCalendarList
qml: [logger:debug] accessToken ////REMOVED_TOKEN////
qml: HTTP 0 Headers:

qml: [logger:debug] fetchGCalCalendars.response.data
qml: XMLHttpRequest.onerror 0
qml: [logger:debug] fetchGCalCalendars.response.data null
```

```
qml: [logger:debug] updateCalendarList
qml: [logger:debug] accessToken ////REMOVED_TOKEN////
kdeinit5: Got EXEC_NEW '/usr/lib/x86_64-linux-gnu/qt5/plugins/kf5/kio/http.so' from launcher.
kdeinit5: preparing to launch '/usr/lib/x86_64-linux-gnu/qt5/plugins/kf5/kio/http.so'
qml: HTTP 0 Headers:

qml: fetchGoogleTasklistList.response HTTP Error 0  0
qml: [logger:debug] fetchGoogleTasklistList.response.data
qml: XMLHttpRequest.onerror 0
qml: fetchGoogleTasklistList.response HTTP Error 0 null 0
qml: [logger:debug] fetchGoogleTasklistList.response.data null
qml: HTTP 0 Headers:

qml: [logger:debug] fetchGCalCalendars.response.data
qml: XMLHttpRequest.onerror 0
qml: [logger:debug] fetchGCalCalendars.response.data null
qml: HTTP 0 Headers:

qml: [logger:debug] fetchGCalCalendars.response.data
qml: XMLHttpRequest.onerror 0
qml: [logger:debug] fetchGCalCalendars.response.data null
```

**Comment by kanocz** [comment]
@TLPNull sorry, but I have no idea... this can be affected by many things like proxy settings or something else... and I really has almost no idea about how this thing works :) I've just fixed login and haven't examined whole code to the deep...

**Comment by geraldosimiao** [comment]
@kanocz Thank you verymuch. Your fork with oauth-helper buildin fixed the problems for me too and I'm very happy to use again eventcalendar synched with my google calendars.

![Screenshot_20230509_234043](https://github.com/Zren/plasma-applet-eventcalendar/assets/72357821/c9df3d75-519d-4fd3-893b-a1ff70a1615e)

**Comment by blackPantherOS** [comment]
Today, with latest git (version 76):

```
"400 error code: invalid_request
The out-of-band (OOB) flow has been blocked in 
order to keep users secure. Follow the Out-of-Band (OOB) 
flow migration guide linked in the developer docs below 
to migrate your app to an alternative method.
Data: redirect_uri=urn:ietf:wg:oauth:2.0:oob"

```

**Comment by kanocz** [comment]
> Today, with latest git (version 76):
> 
> ```
> "400 error code: invalid_request
> The out-of-band (OOB) flow has been blocked in 
> order to keep users secure. Follow the Out-of-Band (OOB) 
> flow migration guide linked in the developer docs below 
> to migrate your app to an alternative method.
> Data: redirect_uri=urn:ietf:wg:oauth:2.0:oob"
> ```

yes, because repo owner hasn't merged any version of fix yet

**Comment by sasa-tomic** [comment]
You can try installing from https://github.com/sasa-tomic/plasma-applet-eventcalendar
